driver-core/src/test/java/com/datastax/driver/core/CCMTestsSupport.java

```diff
@@ -283,31 +284,88 @@ private boolean dirtiesContext() {
+    private static class CCMTestContextKey {
+
+        private final CCMBridge.Builder builder;
+
+        private final int weight;
+
+        private CCMTestContextKey(CCMBridge.Builder builder, int weight) {
+            this.builder = builder;
+            this.weight = weight;
+        }
+    }
+
+    private static class CCMTestContext implements Runnable {
+
+        private CCMBridge ccm;
+
+        private final AtomicInteger refCount = new AtomicInteger(1);
+
+        private volatile boolean evicted = false;
+
+        private CCMTestContext(CCMBridge ccm) {
+            this(ccm, false);
+        }
+
+        public CCMTestContext(CCMBridge ccm, boolean evicted) {
+            this.ccm = ccm;
+            this.evicted = evicted;
+        }
+
+        @Override
+        public void run() {
+            if (ccm != null)
+                ccm.close();
+            ccm = null;
+        }
+
+        private void markEvicted() {
+            evicted = true;
+        }
+
+        private void maybeClose() {
+            if (refCount.get() <= 0 && evicted)
+                POOL.execute(this);
+        }
     }
 
-    private static class CCMTestContextLoader extends CacheLoader<CCMBridge.Builder, CCMBridge> {
+    private static class CCMTestContextLoader extends CacheLoader<CCMTestContextKey, CCMTestContext> {
 
         @Override
-        public CCMBridge load(CCMBridge.Builder builder) {
-            return builder.build();
+        public CCMTestContext load(CCMTestContextKey key) {
+            return new CCMTestContext(key.builder.build());
         }
 
     }
 
-    private static class CCMTestContextWeigher implements Weigher<CCMBridge.Builder, CCMBridge> {
+    private static class CCMTestContextWeigher implements Weigher<CCMTestContextKey, CCMTestContext> {
+
         @Override
-        public int weigh(CCMBridge.Builder key, CCMBridge value) {
-            return value.getInitialNumberOfNodes();
+        public int weigh(CCMTestContextKey key, CCMTestContext value) {
+            return key.weight;
         }
+
     }
 
-    private static class CCMTestContextRemovalListener implements RemovalListener<CCMBridge.Builder, CCMBridge> {
+    private static class CCMTestContextRemovalListener implements RemovalListener<CCMTestContextKey, CCMTestContext> {
 
         @Override
-        public void onRemoval(RemovalNotification<CCMBridge.Builder, CCMBridge> notification) {
-            if (notification.getValue() != null) {
-                LOGGER.debug("Evicting: " + notification.getValue());
-                notification.getValue().close();
+        public void onRemoval(RemovalNotification<CCMTestContextKey, CCMTestContext> notification) {
+            CCMTestContext context = notification.getValue();
+            if (context != null && context.ccm != null) {
+                LOGGER.debug("Evicting: {} because of {}", context.ccm, notification.getCause());
+                context.markEvicted();
+                context.maybeClose();
```

Comment:
This situation happens specially when cache maximum size is low: immediately after creating a CCMBridge instance, if it's weight is greater than the total maximum cache weight, it is immediately evicted:

```
Executing: ccm create test_40 -i 127.0.1. -v 2.1.9 -n 3 -p RandomPartitioner --vnodes --config-dir=/var/folders/z4/z9k05zy53bs00rxf_b52tfth0000gn/T/1452856191764-0
Evicting: CCM cluster test_40 because of SIZE
Using newly-created CCM cluster test_40
```

If the cluster is closed at that moment, the test will fail.